pkg/api/mock/mock_api_implementation_test.go

```diff
@@ -0,0 +1,226 @@
+			&clouddnsv1.Record{Name: "txt-record", Type: "TXT", RData: "hello world"},
+		}
+
+		objects := []types.Object{
+			&clouddnsv1.Zone{Name: "mock.go-anxcloud.test"},
+			testVLAN,
+		}
+
+		objects = append(objects, dnsRecords...)
+
+		It("can Create resources", func() {
+			for _, o := range objects {
+				err := a.Create(context.TODO(), o)
+				Expect(err).ToNot(HaveOccurred())
+			}
+		})
+
+		It("supports not found errors on Get operations", func() {
+			err := a.Get(context.TODO(), &vlanv1.VLAN{Identifier: "nonexistent-identifier"})
+			Expect(err).To(MatchError(api.ErrNotFound))
+		})
+
+		It("can Get resources", func() {
+			vlan := &vlanv1.VLAN{Identifier: testVLAN.Identifier}
+			err := a.Get(context.TODO(), vlan)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(testVLAN).To(Equal(vlan))
+		})
+
+		It("can Update resources", func() {
+			copy, err := copystructure.Copy(testVLAN)
+			Expect(err).ToNot(HaveOccurred())
+
+			vlan := copy.(*vlanv1.VLAN)
+			vlan.Name = "def"
+
+			err = a.Update(context.TODO(), vlan)
+			Expect(err).ToNot(HaveOccurred())
+
+			searchVLAN := &vlanv1.VLAN{Identifier: testVLAN.Identifier}
+			err = a.Get(context.TODO(), searchVLAN)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(searchVLAN).To(Equal(vlan))
+			Expect(searchVLAN).To(Not(Equal(testVLAN)))
+		})
+
+		It("returns ErrNotFound when object to Update not existing", func() {
+			err := a.Update(context.TODO(), &vlanv1.VLAN{Identifier: "nonexistent-identifier"})
+			Expect(err).To(MatchError(api.ErrNotFound))
+		})
+
+		It("can List resources [ObjectChannel]", func() {
+			var oc types.ObjectChannel
+			err := a.List(context.TODO(), &clouddnsv1.Record{}, api.ObjectChannel(&oc))
+			Expect(err).ToNot(HaveOccurred())
+
+			var res []types.Object
+			for retriever := range oc {
+				var obj clouddnsv1.Record
+				err := retriever(&obj)
+				Expect(err).ToNot(HaveOccurred())
+				res = append(res, &obj)
+			}
+
+			Expect(res).To(ConsistOf(dnsRecords))
+		})
+
+		It("can List resources [Paged]", func() {
+			var pi types.PageInfo
+			err := a.List(context.TODO(), &clouddnsv1.Record{}, api.Paged(1, 1, &pi))
+			Expect(err).ToNot(HaveOccurred())
+
+			res := make([]types.Object, 0, pi.TotalItems())
+			var page []types.Object
+			for pi.Next(&page) {
+				res = append(res, page...)
+			}
+
+			Expect(res).To(ConsistOf(dnsRecords))
+		})
+
+		It("can Destroy resources", func() {
+			err := a.Destroy(context.TODO(), &vlanv1.VLAN{Identifier: testVLAN.Identifier})
+			Expect(err).ToNot(HaveOccurred())
+
+			err = a.Get(context.TODO(), &vlanv1.VLAN{Identifier: testVLAN.Identifier})
+			Expect(err).To(MatchError(api.ErrNotFound))
+		})
+
+		It("returns ErrNotFound when object to Destroy not existing", func() {
+			err := a.Destroy(context.TODO(), &vlanv1.VLAN{Identifier: "nonexistent-identifier"})
+			Expect(err).To(MatchError(api.ErrNotFound))
+		})
+
+	})
+
+	Context("Custom gomega matcher", Ordered, func() {
```

Comment:
any chance to extract some of those into example tests to have them in the docs?